issue-40-built-in-benchmarking-tool.md

```diff
@@ -0,0 +1,32 @@
+## Add Built-in Benchmarking Tool
+
+## Summary
+
+There is a need to support a built-in benchmarking tool in Submariner.
+This is to provide users with easy way to benchmark the network performance within and across clusters, and experiment with different tuning of the system.
+
+## Proposal
+
+Add a new ability to have a built-in benchmark tool.
+Implement a new Subctl command that will use iperf3 internally to benchmark the throughput between the two nodes.
+
+scheduling server and client can be placed within the same cluster or spread across multiple clusters.
```

Comment:
I agree that the main use case we should target is testing *across* different clusters.. but technically one should be able to use the tool as is to also benchmark within a single cluster. Or do you suggest we explicitly block this option?

Reply:
While measuring Submariner's network performance, we found it useful to have some sort of baseline within a cluster to compare the cross-cluster traffic to. For example, if intercluster network performance is low, it makes sense that intracluster network performance is also limited. So it's not an exact comparison, but it does provide a useful context for the Submariner benchmark.

Reply:
Thanks for the input @greenren 

I feel that this tool's strong point would be to measure the metrics across clusters, hence to me that should be the focus.
Obviously you can measure intra-cluster traffic also, but as submariner is not involved in the networking there (we strive to be as agnostic as possible about the specific cluster CNI in fact), it doesnt make much sense for these measurements to be a focus for this tool.

How I see this tool being useful is try out different submariner configurations across clusters and measuring and comparing that, for example to check submariner with and without globalnet and compare the impact of using that feature on the data plane. Another comparison is to check different cable drivers or cable driver configurations.

While it's true that intra-cluster network performance would affect the inter-cluster results, it would do so in a consistent matter which is comparable (if the disturbance isn't transient), but comparing inter-cluster results to intra-cluster results would give no benefit as it's akin to comparing the speed of a virtual NIC to the speed of a physical NIC.

To be clear, I'm not saying this tool couldn't be possibly used to measure intra-cluster performance, I just think that as it's not the focus nor the intention of this tool, this mode should not be advertised or supported. I.E. it would be provided "as is" without any guarantee to serve this purpose.

I'm sure there are probably many networking tools that can be used to measure intra-cluster networking performance, possibly even in the K8s ecosystem.

Reply:
I see @mkolesnik point here about... not measuring something that it's specific to submariner.

But since supporting intra-cluster should not make it specifically much more complicated (it adds the intra-node / node to node dimension) .. I think it's good to have this so we can provide a good baseline to compare, which proved very useful as a baseline as @greenren said.

May be this is something we can propose as a second iteration (given that it requires extra work, although not much). ?

Reply:
I agree with @greenren and @mangelajo. Since Submariner also programs certain iptable chains/routing rules on the nodes, it will be helpful if we allow the tool to validate throughput within the local cluster. This way, we can use the tool to compare the throughput with and without Submariner in the local cluster as well.